proto/substrait/algebra.proto

```diff
@@ -58,7 +58,14 @@ message RelCommon {
 // The scan operator of base data (physical or virtual), including filtering and projection.
 message ReadRel {
   RelCommon common = 1;
-  NamedStruct base_schema = 2;
+  repeated Field fields = 2;
+  message Field {
+    Type type = 1;
```

Comment:
I don't think this pattern works. The problem is that you lose the ability to express custom metadata and any other than top level of fields. For example, if I have:

Imagine the following named struct:

```
schema = struct<i32, struct<i64,double>>
field names = myI32, myStruct, myI64, myDouble.
```

Using the proposal here, there is now way to add metadata to bar or baz. We need to come up with a better proposal. My inclination is that we should extension support to types for custom information.

Reply:
Hi Jacques, do you refer to below case? I think it is OK because metadata is a repeated message.
```
schema = struct<i32, struct<i64,double>>
field names = myI32, myStruct, myI64, myDouble
metadatas = i32Meta, structMeta, i64Meta, doubleMeta
```
Would the extension to type affect too many usages? Here we mainly want to add some metadata for table scanning.

Reply:
hi @jacques-n, this change aims to allow custom meta data for a specific column especially for table reading or writing. E.g. compression, dictionary encoding, partition (as well as the partition key format maybe), etc. In this way, one can describe more details about one column besides the name and type. Would you like to take a look if it can work as above way if I understand correctly?

Reply:
Thanks for your comment. I tried the similar thing but looks like current way is more preferred according to previous comments. For one thing, the metadata can be duplicated with advanced_extension, for another, the named struct seems to be something to be refactored. For details, maybe you can check https://github.com/substrait-io/substrait/pull/298#issuecomment-1262320376. From my personal opinion, adding metadata to Field enables more information for a specific column, and advanced_extension can be used to carry more information for a ReadRel.